typescript/sdk/src/core/adapters/SealevelCoreAdapter.ts

```diff
@@ -0,0 +1,149 @@
+import { PublicKey } from '@solana/web3.js';
+
+import { HexString, pollAsync } from '@hyperlane-xyz/utils';
+
+import { BaseSealevelAdapter } from '../../app/MultiProtocolApp';
+import {
+  ProviderType,
+  TypedTransactionReceipt,
+} from '../../providers/ProviderType';
+import { ChainName } from '../../types';
+import { CoreAddresses } from '../contracts';
+
+import { ICoreAdapter } from './types';
+
+// https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/sealevel/programs/mailbox/src/processor.rs
+const MESSAGE_DISPATCH_LOG_REGEX = /Dispatched message to (.*), ID (.*)/;
+
+// This adapter just routes to the HyperlaneCore
+// Which implements the needed functionality for EVM chains
+export class SealevelCoreAdapter
+  extends BaseSealevelAdapter<CoreAddresses>
+  implements ICoreAdapter
+{
+  extractMessageIds(
+    sourceTx: TypedTransactionReceipt,
+  ): Array<{ messageId: HexString; destination: ChainName }> {
+    if (sourceTx.type !== ProviderType.SolanaWeb3) {
+      throw new Error(
+        `Unsupported provider type for SealevelCoreAdapter ${sourceTx.type}`,
+      );
+    }
+    const receipt = sourceTx.receipt;
+    const logs = receipt.meta?.logMessages;
+    if (!logs)
+      throw new Error('Transaction logs required to check message delivery');
+    const parsedLogs = SealevelCoreAdapter.parseMessageDispatchLogs(logs);
+    if (!parsedLogs.length) throw new Error('Message dispatch log not found');
+    return parsedLogs.map(({ destination, messageId }) => ({
+      messageId: Buffer.from(messageId, 'hex').toString('hex'),
+      destination: this.multiProvider.getChainName(destination),
+    }));
+  }
+
+  async waitForMessageProcessed(
+    messageId: string,
+    destination: ChainName,
+    delayMs?: number,
+    maxAttempts?: number,
+  ): Promise<void> {
+    const destinationMailbox =
+      this.multiProvider.getChainMetadata(destination).mailbox;
+    const pda = SealevelCoreAdapter.deriveMailboxMessageProcessedPda(
+      destinationMailbox,
+      messageId,
+    );
+    const connection = this.multiProvider.getSolanaWeb3Provider(destination);
+
+    await pollAsync(
+      async () => {
+        // If the PDA exists, then the message has been processed
+        // Checking existence by querying for balance
+        await connection.getBalance(pda);
```

Comment:
does this throw if an account has no balance? Fwiw I tried it locally and it just returned 0 for a random account without throwing

I can't think why anyone would want to do this but because anyone can transfer lamports to whatever account they want, just relying upon balance as a way to see if the PDA "exists" has the edge case of someone sending it some dust

Wdyt of https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getAccountInfo, which returns null if an account has no balance and no data, and having the condition be `accountInfo !== null && accountInfo.data.length > 0`. The real condition of a PDA "existing" is more that there's some data there, which can only be put there by the program when it processes a message